Support nested context paths.

chickenchickenlove opened the pull request:
### Motivation:
- Currently, `armeria` supports only `1-depth` context paths. Sometimes, user want deeper context paths than 1-depth when they use `contextPaths()`.

### Modifications:
- Make `ContextPathServiceBuilder` tree to support nested context paths. 
- Add public functions. 
  - `before()` is for going back previous node.
  - `contextPath()` is for adding context paths and making child `ContextPathServiceBuilder`
- Add package-private and private functions.
  - `parent()`: To give Child `contextPathServiceBuilder` parent object. because of this, child can return parent object when `and()` is called.
  - `virtualHostbuilder()`: to relay their context via `ContextPathServiceBuilder` tree.
  - `mergeContextPaths()`: to merge previous context paths and current context paths. 

### Example
- I make a example armeria codes and test codes
- repository : https://github.com/chickenchickenlove/nestedcontext
- armeria example : https://github.com/chickenchickenlove/nestedcontext/blob/master/src/main/java/org/example/Main.java
- test example: https://github.com/chickenchickenlove/nestedcontext/blob/master/pytest/my_test.py


### Result:
- Closes https://github.com/line/armeria/issues/5758
- User can use nested context paths. for example, 
```java
sb.contextPath("/rest")
      .contextPath("/catalog")
          .service("/product", new GetProductService())
          .service("/products", new ProductsHandler())
          .before()
      .contextPath("/cart")
          .contextPath("/foo")
               .contextPath("/bar")
                    .service("/checkout", new CheckoutService());
                    .and()
  .contextPath("/gql")
      .service("/catalog", new GraphQLService());
```

`/rest/catalog/product` => `getProductService`
`/rest/catalog/products` => `productsHandler`
`/rest/cart/foo/bar/checkout` => `checkoutService`
`/gql/catalog` => `GraphQLService`

chickenchickenlove commented:
I would like to receive feedback on whether this PR is heading in the right direction, as it is currently a draft.

After fix direction, I will write unit tests and Java docs based on the BaseContextPathsTest code.

When you have time, please take a look 🙇‍♂️

ikhoon commented:
I didn't see that a fluent style is good for nested context paths in terms of API design.
How about taking a lambda expression as a second parameter to leverage Kotlin trailing lamdas?
https://kotlinlang.org/docs/lambdas.html#passing-trailing-lambdas

I imagined the following style API. We used `contextPath()` at the top level but I am not sure if the naming is also good for the nested routes. 
```kt
Server
  .builder()
  .contextPath("/rest") {
     contextPath("/catalog") {
        service("/product", new GetProductService())
        service("/products", new ProductsHandler())/
     }
  }
  .contextPath("/cart") {
     contextPath("/foo") {
        contextPath("/bar") {
           service("/checkout", new CheckoutService())
        }
     }
  }
  .contextPath("/gql") {
      service("/catalog", new GraphQLService())
   }
```
Related work: https://ktor.io/docs/server-routing.html#nested_routes

chickenchickenlove commented:
@ikhoon nim, Thanks for your guidelines.
I didn't know about `KTOR`. i will take a look and follow how `KTOR` supports nested context path.

chickenchickenlove commented:
@ikhoon nim, i change code style from fluent style to nested lambda expression.
For example, we can use feature of nested context like this.
```java
        sb.baseContextPath("/api")
          .contextPath("/context-path/a1", "/context-path/a2")
          .nestedContext()
          .contextPaths(Set.of("/b1", "/b2"), ctx1 -> ctx1
                  .annotatedService(new Object() { ... })
                  .service("/my-service", new HttpService() { ... })
                  .contextPaths(Set.of("/c1", "/c2"), ctx2 -> ctx2
                          .service("/my-service1", new HttpService() { ... })
          .contextPaths(Set.of("/b3", "/b4"), ctx11 -> ctx11
                  .service("/my-service", new HttpService() { ... });

       sb.virtualHost("foo.com")
          .contextPath("/virtual-foo")
          .nestedContext()
          .contextPaths(Set.of("/a1", "/a2"), ctx -> ctx
                  .service("/my-service1", new HttpService() { ... }));
```

To get detail code, you can refer to this [java code](https://github.com/chickenchickenlove/nestedcontext/blob/master/src/main/java/org/example/Main.java#L16-L152). 
Also, i create [pytest code](https://github.com/chickenchickenlove/nestedcontext/blob/master/pytest/my_test.py#L5-L69) to test whether endpoint is valid, corresponding to java code above.

When you have time, please take another look. 🙇‍♂️

chickenchickenlove commented:
@ikhoon nim, gently ping. 
When you have time, please take a look 🙇‍♂️ 
Or would you prefer to not proceed with this PR any further, in which case I'd be happy to close this PR.

ikhoon commented:
> ```java
>  sb.baseContextPath("/api")
>    .contextPath("/context-path/a1", "/context-path/a2")
>    .nestedContext()
>    .contextPaths(Set.of("/b1", "/b2"), ctx1 -> ctx1  // ctx1 scope start (lambda function)
> ```

Is it possible to build nested context paths without using `.nestedContext()`? I would like the API to be designed in a form close to a DSL.

chickenchickenlove commented:
@ikhoon nim, sure.
I fixed the codes. 
Two styles are supported.

I updated [example codes ](https://github.com/chickenchickenlove/nestedcontext/blob/master/src/main/java/org/example/Main.java)and [API test codes.](https://github.com/chickenchickenlove/nestedcontext/blob/master/pytest/my_test.py).

### Style1
It starts with `contextPath(Iterable<String> paths)`.

```java
      sb.baseContextPath("/api")
        .contextPath("/context-path/a1", "/context-path/a2")
        .contextPaths(Set.of("/b1", "/b2"), ctx1 -> ctx1  // ctx1 scope start (lambda function)
                .annotatedService(new Object() { ... })
                .service("/my-service", new HttpService() { ... })
                .contextPaths(Set.of("/c1", "/c2"), ctx2 -> ctx2 // ctx2 scope start (lambda function)
                        .service("/my-service1", new HttpService() { ... })
                        .contextPaths(Set.of(...), ctx3 -> ctx3  // ctx3 scope start (lambda function)
                                 .service("/...", new HttpService() { ... })
                                 ) // ctx3 scope end.
                       ) // ctx2 scope end.                   
         ) // ctx1 scope end.
        .contextPaths(Set.of("/b3", "/b4"), ctx11 -> ctx11
                .service("/my-service", new HttpService() { ... });
```

### Style2
It starts with `contextPath(Iterable<String> paths, Consumer<ContextPathServicesBuilder> context)`.

```java
      sb.baseContextPath("/api")
        .contextPath(Set.of("/b1", "/b2"), ctx1 -> ctx1  // ctx1 scope start (lambda function)
                .annotatedService(new Object() { ... })
                .service("/my-service", new HttpService() { ... })
                .contextPaths(Set.of("/c1", "/c2"), ctx2 -> ctx2 // ctx2 scope start (lambda function)
                        .service("/my-service1", new HttpService() { ... })
                        .contextPaths(Set.of(...), ctx3 -> ctx3  // ctx3 scope start (lambda function)
                                 .service("/...", new HttpService() { ... })
                                 ) // ctx3 scope end.
                       ) // ctx2 scope end.                   
         ) // ctx1 scope end.
        .contextPaths(Set.of("/b3", "/b4"), ctx11 -> ctx11
                .service("/my-service", new HttpService() { ... });
```

When you have time, PTAL 🙇‍♂️

ikhoon commented:
Could you explain the difference between `contextPath()` and `contextPaths()`? The method names alone don't make it clear.

chickenchickenlove commented on core/src/main/java/com/linecorp/armeria/server/ServerBuilder.java:
```diff
@@ -1229,6 +1229,16 @@ public ContextPathServicesBuilder contextPath(Iterable<String> contextPaths) {
+     * @return
+     */
+    @UnstableApi
+    public ContextPathServicesBuilder toContextBuilder() {
```
It works. applied.

chickenchickenlove commented:
Hi, @ikhoon nim.
Sorry to bother you.
When you have time, could you take a look?

chickenchickenlove commented:
@ikhoon , HI! 
Thanks for your comments.
I made new commits based on your comments. (Add test case, Java Docs) 
Also, I found the defects in `RejectRouterHandler.FAIL` and issued it (#6498) and made a patch(#6499)

If #6499 is merged, I can write test code for testing duplicated routing rules in nested context path. 

When you have time, please take another look. 🙇‍♂️

ikhoon commented on core/src/main/java/com/linecorp/armeria/server/ContextPathServicesBuilder.java:
```diff
@@ -100,4 +100,15 @@ public ContextPathServicesBuilder annotatedService(
+
+    @Override
+    public ContextPathServicesBuilder contextPath(Iterable<String> paths,
+                                                  Consumer<ContextPathServicesBuilder> customizer) {
```
> At the same time, I believe keeping contextPath(Iterable<String> paths, ...) is useful for common use cases, especially API versioning.

It makes sense. Let's support both of them.

ikhoon commented on core/src/main/java/com/linecorp/armeria/server/ContextPathServicesBuilder.java:
```diff
@@ -100,4 +102,22 @@ public ContextPathServicesBuilder annotatedService(
+    @Override
+    public ContextPathServicesBuilder contextPath(String path,
+                                                  Consumer<ContextPathServicesBuilder> customizer) {
+        contextPath(ImmutableSet.of(path), customizer);
```
Should we check `requiresNonNull` for both arguments?

chickenchickenlove replied:
Addressed!

ikhoon commented on core/src/test/java/com/linecorp/armeria/server/HttpServerNestedContextPathTest.java:
```diff
@@ -0,0 +1,498 @@
+                      });
+                  });
+              })
+              .contextPath("/single-path1", ctx1 -> {
```
Optional) It would be nice if we had some Kotlin tests.

chickenchickenlove replied:
OK, Let me try!

> If https://github.com/line/armeria/pull/6499 is merged, I can write test code for testing duplicated routing rules in nested context path.

I will create bunch of test codes to validate duplicate nested context path. 
At that time, I will create some kotlin tests as well!

chickenchickenlove commented on core/src/main/java/com/linecorp/armeria/server/AbstractContextPathServicesBuilder.java:
```diff
@@ -409,4 +411,61 @@ public T and() {
+        final Set<String> mergedContextPaths = new HashSet<>();
+        for (String currentContextPath : contextPaths()) {
+            for (String childContextPath : paths) {
+                RouteUtil.ensureAbsolutePath(childContextPath, "contextPath");
```
@ikhoon Thanks for your comments! 
IMHO, It is not enough, so I think we must validate the paths before they are merged. 
If we don't validate the context path before they are merged,  
the codes below will not be failed. 
`Builder` will see `/hellorelative` and it don't violate absolute path.
```java
    @Test
    void relativeNestedContextPathShouldBeFailed() {
        final ServerBuilder serverBuilder = new ServerBuilder().baseContextPath("/api");
        assertThatThrownBy(() -> {
            serverBuilder.contextPath(ImmutableSet.of("/hello"), ctx1 -> {
                ctx1.contextPath(ImmutableSet.of("relative"), ctx2 -> {
                    ctx2.serviceUnder("/hello", (ctx, req) -> HttpResponse.of(HttpStatus.OK));
                });
            });
        }).isInstanceOf(IllegalArgumentException.class);
    }
```

So, I think keeping `RouteUtil.ensureAbsolutePath(...)` in `mergeContextPaths(...)` is reasonable.
What do you think?

ikhoon commented on core/src/main/java/com/linecorp/armeria/server/ServerBuilder.java:
```diff
@@ -1345,6 +1345,44 @@ public ContextPathServicesBuilder contextPath(Iterable<String> contextPaths) {
+    /**
+     * Applies the specified {@code customizer} to a {@link ContextPathServicesBuilder} rooted at the
+     * given {@code contextPaths} of the default {@link VirtualHost}.
+     * @param contextPaths the context paths to group services under; must be non-empty and absolute
```
>  must be non-empty 

This constrain doesn't seem to be applied.

minwoox commented on core/src/main/java/com/linecorp/armeria/server/AbstractContextPathServicesBuilder.java:
```diff
@@ -409,4 +411,60 @@ public T and() {
     }
+
+    VirtualHostBuilder virtualHostBuilder() {
+        return this.virtualHostBuilder;
```
nit:
```suggestion
        return virtualHostBuilder;
```

chickenchickenlove replied:
Addressed!
